Setup Certificate Search

king-alexander opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->
This pull request makes several changes to setup the [sslmate Certificate Search API](https://sslmate.com/ct_search_api/) as our method to monitor and query Certificate Transparency logs.
## 💭 Motivation and context ##

<!-- Why is this change required? -->
The current service we use to monitor Certificate Transparency logs is too unreliable. This change provides more stability at the scale we operate.

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
Over a few dry runs of `load_certs.py`, I have processed approximately 200 domains. 

In addition, all unit and system tests passed (`Python 3.10.4`, `pytest-7.2.0`).
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->
<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been approved. -->

- [x] Revert dependencies to default branches.
- [x] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Add a tag or create a release.

mcdonnnj committed:
Add an ownership comment to the Dependabot configuration

Add a comment that states that the commented out ignore directives are
managed by cisagov/skeleton-generic.

mcdonnnj committed:
Add a CodeQL status badge to the README

king-alexander committed:
Create the API Key secret

This commit makes the sslmate Certificate Search API Key a secret and
grants access to admiral services.

king-alexander committed:
Make the earliest expired date timezone aware

In order to compare the certificate's expiration date with the earliest
expired date, both must be timezone aware datetime objects. This change
localizes the earliest expired date to UTC.

king-alexander committed:
Retrieve new log issuances

The Certificate Search API exposes issuance objects, which consolidate
records across multiple Certificate Transparency logs. You cannot query
these issuances by id, so instead we must generate a sequence of new,
unique issuances that contain the certificates we need to add to our
database.

king-alexander committed:
Convert DER-encoded certificate to PEM-encoded certificate

king-alexander committed:
Remove default rate limit for tasks

king-alexander committed:
Set task rate limit

The Small Tier of the sslmate Cert Search API permits up to 10
full-domain queries per hour.

king-alexander committed:
Remove expired parameter

The sslmate Cert Search API does not define an expired parameter. These
changes remove the expired parameter from the summary_by_domain function
signature and the local expired variable passed to summary_by_domain in
the load_certs script.

king-alexander committed:
Redefine cert_by_id

The work this task needs complete is significantly lighter using the
sslmate Cert Search API. Fetching a certificate's information does not
require a separate query. Instead, the task returns the certificate
associated with a particular issuance and logs the id.

jsf9k commented on examples/load_certs.py:
```diff
@@ -52,29 +60,29 @@ def cert_id_exists_in_database(log_id):
     duplicate_log_ids = set()
     for i in tqdm(cert_list, desc="Subjects", unit="entries", leave=False):
         log_id = i["id"]
+        name = i["dns_names"][0]  # take the first dns name available
```
Is there value in keeping _all_ DNS names?

king-alexander replied:
See [783d7ee](https://github.com/cisagov/admiral/pull/31/commits/783d7ee634bbee8d3da93176896822e9f04e9802).

jsf9k replied:
I'm fine with that comment.

jsf9k commented on src/admiral/certs/tasks.py:
```diff
@@ -28,63 +28,53 @@
     req = requests.get(
         url,
-        headers={"User-Agent": "cyhy/2.0.0"},
+        headers={"Authorization": f"Bearer {key}", "User-Agent": "cyhy/2.0.0"},
```
Do we want the user-agent to be something more specific than `cyhy`?  Maybe something like `admiral/1.2.3` would be better, since we can automatically update the string with the appropriate version.  It isn't clear to me where the `2.0.0` version for `cyhy` is coming from.

mcdonnnj replied:
I think the `cyhy/2.0.0` version was based on this being something of a demo for an envisioned CyHy v2 rework from long in the past.

dav3r replied:
I like @jsf9k's suggestion of using the admiral version here.

king-alexander replied:
See [f01a4a9](https://github.com/cisagov/admiral/pull/31/commits/f01a4a9387878a44186ade1ee09da43e17484935).

jsf9k replied:
Nice!

jsf9k reviewed (commented):
Nice work!

I have a couple of comments/thoughts that will require some input from @cisagov/team-ois.

dav3r commented on examples/load_certs.py:
```diff
@@ -34,7 +38,11 @@
 # Globals
 PP = pprint.PrettyPrinter(indent=4)
-EARLIEST_EXPIRED_DATE = parser.parse("2018-10-01")
+EARLIEST_EXPIRED_DATE = default_tzinfo(
```
I know it's outside of the scope of this PR, but can you please add a comment explaining where this date came from and if we would ever need to update it to a different date?

Kudos for making this timezone-aware though!

king-alexander replied:
See [53ce68f](https://github.com/cisagov/admiral/pull/31/commits/53ce68fee03579b630c8e6577e48d8feb3ace9a8).

dav3r replied:
Excellent comment. 👍

dav3r commented on examples/load_certs.py:
```diff
@@ -52,29 +60,29 @@ def cert_id_exists_in_database(log_id):
     duplicate_log_ids = set()
     for i in tqdm(cert_list, desc="Subjects", unit="entries", leave=False):
         log_id = i["id"]
+        name = i["dns_names"][0]  # take the first dns name available
```
We strive for proper capitalization whenever possible:
```suggestion
        name = i["dns_names"][0]  # take the first DNS name available
```

dav3r commented on src/admiral/certs/tasks.py:
```diff
@@ -28,63 +28,53 @@
     retry_backoff=True,
     retry_jitter=True,
     retry_kwargs={"max_retries": 16},
+    rate_limit="10/h",
```
Please alphabetize `rate_limit` here.

king-alexander replied:
See [d6a3729](https://github.com/cisagov/admiral/pull/31/commits/d6a37296c4a20ab0e8b66612d80165eda759d4d1).

dav3r commented on src/admiral/certs/tasks.py:
```diff
@@ -28,63 +28,53 @@
 
-    wildcard_param = "%." if subdomains else ""
-    expired_param = "" if expired else "&exclude=expired"
+    # fetch api key
```
```suggestion
    # read SSLMate API key
```

dav3r commented on src/admiral/certs/tasks.py:
```diff
@@ -28,63 +28,53 @@
-    expired_param = "" if expired else "&exclude=expired"
+    # fetch api key
+    key = ""
+    with open("/run/secrets/sslmate-api-key.txt", "r") as file:
```
"r" is the default mode, so it can be removed here:
```suggestion
    with open("/run/secrets/sslmate-api-key.txt") as file:
```

dav3r commented on src/admiral/certs/tasks.py:
```diff
@@ -28,63 +28,53 @@
-    logger.info(f"Fetching cert data from CT log for id: {id}.")
+@shared_task
+def cert_by_id(issuance):
+    """Fetch a certificate by log ID.
```
The name and docstring description here no longer match what is being done in this function.  Maybe rename it to `cert_by_issuance` or `cert_from_issuance`?

king-alexander replied:
See [7ea3b73](https://github.com/cisagov/admiral/pull/31/commits/7ea3b73e25ad32c4c38178b696bd71e0b477db90).

dav3r reviewed (changes requested):
Please take a look at my feedback and let me know if you have any questions.

king-alexander committed:
Capitalize "DNS"

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

king-alexander committed:
Update API key comment

This change better describes the action that occurs. The API key isn't fetched; it is read from a file.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

king-alexander committed:
Remove the mode argument from file open

Read mode ("r") is assumed when the mode argument is omitted.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

king-alexander committed:
Alphabetize summary_by_domain task options

king-alexander committed:
Rename "cert_by_id" task to "cert_by_issuance"

The name and docstring no longer matched the work undertaken by the
function. Renaming the task "cert_by_issuance" is a more accurate
description.

king-alexander committed:
Explain the choice of earliest expired date

king-alexander committed:
Specify admiral version in the user-agent string

king-alexander committed:
Explain use of DNS names

dav3r reviewed (approved):
Nice work here @king-alexander 👍

mcdonnnj committed:
Remove LGTM badges from the README

Now that LGTM has been deprecated in favor of GitHub code scanning
(CodeQL) we have removed the integration from the organization. As a
result we need to remove these badges.

mcdonnnj committed:
Bump version from 0.1.0 to 0.2.0

mcdonnnj committed:
Add ignore directives to the dependabot configuration

Add ignore directives and an attribution comment to the dependabot
configurtation. These will be uncommented in descendants of this
project (cisagov/skeleton-python-library) so that version changes to
these Actions are controlled through this project.
